Supporting embedding models

jc9123 opened the pull request:
Added support for BGE-M3 which is an embedded model. The embedded vector is stored in the .embed of the returned Request Object. A example test case can be found in tests/models/test_embedded.py.

FIX #9847

jc9123 committed:
added OpenAI embedded model API call

jc9123 committed:
added openai embedded api call

jc9123 committed:
added openai embedded api call

jc9123 committed:
save

jc9123 committed:
added embedded model

jc9123 committed:
Merge branch 'main' into supporting_embedded_model

robertgshaw2-neuralmagic commented:
Thank you for the PR.

A couple comments:

- The major thing that should be done to add BGE to vLLM is that the model should be re-implemented using the layers from vLLM. vLLM has optimized inference kernels for most layers in transformers, we should use these in the implementation. Ideally, you would also use the `ColumnParallel` and `RowParallel` layers to support multi-gpu inference as well. Currently, it looks like you are using the transformers model inside vLLM in the definition of BGE. Take a look at the implementation of some of the other models for inspiration. You will need to modify attention to be bidirectional, but they should be instructive

- I do not think that we should be adding a new input `params_dict` to vLLM. Additionally, we should not be tokenizing the input inside the forward method. The inputs to encoder models are tokens, just like for LLMs. We should be able to leverage the existing tokenization / sequence group / input_metadata infrastructure for encoder models

CatherineSue commented:
@robertgshaw2-neuralmagic KV cache is unnecessary for embedding generation. Is there any plan to disable KV cache to better support embedding models?

jc9123 committed:
fixed tokenizing and cleaned up code for bgem3

jc9123 committed:
removed prompt dict

jc9123 committed:
Merge remote-tracking branch 'upstream/main' into add_embedded_model

jc9123 committed:
fixed merge conflict

jc9123 commented:
I've removed param_dict and and replaced it with the existing tokenization infrastructure. I think I've changed the inference layer to use vLLM's inference layer as well, but I may have missed something so let me know if I need to fix anything. Thanks!

robertgshaw2-neuralmagic commented:
@jc9123 i will take a look

robertgshaw2-neuralmagic commented:
I left a few tactical nits above which should be addressed directly.

However, there is significant work to do to get this PR into a place that is ready to merge. I would suggest we split up the work into a series of PRs.

### Step 1: Basic functionality using Hugging Face models
In this step, we will get basic functionality working, meaning:
- Model is implemented using HF transformers
- Create new python front end called `EmbeddingModel` to be used in place of `LLM` for embedding models. This class will be used to receive text input and return embeddings

Additionally, make sure there are checks that turn off incompatible features are turned off for embedding models:
- Tensor parallelism > 1
- Speculative decoding
- Automatic prefix caching
- LORA
- Quantization
- CUDAGraphs

I will work closely with you on this PR to complete Milestone 1.

### Step 2: Implement the model properly
In this step, we will implement the model using the vllm layer primitives rather than using the transformers implementation of the model. This will enable us to support tensor parallelism and have more optimized inference.

The goal is to remove the need to import transformers and instead use the vLLM layer primitives that are optimized for inference like:
- `ColumnParallelLinear`, `RowParallelLinear`, `Attention`, etc.

### Step 3: Add the OpenAI Server Front
- Add support in the `entrypoints/openai` for the embeddings functionality.
- https://platform.openai.com/docs/api-reference/embeddings

### Step 4: Expand feature set [optional]
- Multimodal embedding models
- Support quantized models
- Support quantized embeddings
- Support variable length embeddings
- [ others that I am missing ]

Commit:
added xlm roberta scaffold

Commit:
fixed nits

Commit:
fixed nit again

Commit:
fixed nits again again

Commit:
fixed nits again again again

Commit:
newline on Attention

robertgshaw2-neuralmagic commented:
@jc9123 I made a PR on your branch to:

- fix the nits listed here
- get started implementing the `XLMRobertaModel`

robertgshaw2-neuralmagic commented:
### Milestone 1 Workplan

#### A) Finish the `XLMRobertaModel` definition

In the PR I did against your branch, I setup the basics of the model, which creates the weights using the config during `init` and then loads the weights during `load_weights()` (which is called by vLLM when the model is being loaded.

```python
from transformers import XLMRobertaConfig
from xlm_roberta import XLMRobertaModel

model_id = "BAAI/bge-m3"
config = XLMRobertaConfig.from_pretrained(model_id)
model = XLMRobertaModel(config=config)
model.load_weights(model_id)
```

You need to finish off this function by implementing the `forward` method. `forward` needs to convert from the input data passed around by vllm into the formats expected by the hugging face implementation of the model (which we are using here).

Note: `BAAI/bge-m3` uses the `XLMRobertaModel`. So In vLLM we will implement the `XLMRobertaModel` architecture and then any embedding model that uses this architecture can use the definition

#### B) Update the `LLM` front end UX

I would like the UX to look something like this:

```python
from vllm import LLM, EmbeddingParameters

# LLM should autodetect this is an embedding model since it uses `XLMRobertaModel`
model = LLM("BAAI/bge-m3")
parameters = EmbeddingParameters()
sentences = ["The quick...", "brown fox...", ... "the lazy..."]
model.encode(sentences, parameters=parameters)
```

Then, we can add things like `normalize`, `return_sparse`, `return_dense`, `return_colbert_vecs` via the `EmbeddingParameters` in the future. [ This does not need to be done for this PR ]

We can look at the following repos for potential features to support via `encode`
- `SentenceTransformers` https://www.sbert.net/docs/package_reference/SentenceTransformer.html
- `FlagEmbedding` https://github.com/FlagOpen/FlagEmbedding/tree/master/FlagEmbedding/BGE_M3

jc9123 committed:
Merge pull request #1 from neuralmagic/rs/embedding-step-1

Rs/embedding step 1

jc9123 commented:
Hi Robert, thank you for the detail feedback! I will begin working on milestone 1 and let you know if I have any questions.

robertgshaw2-neuralmagic commented:
@jc9123 updated comment above

jc9123 commented:
> @jc9123 updated comment above

Thanks! I will check in with you after finishing steps a-d.

ywang96 commented on vllm/outputs.py:
```diff
@@ -75,6 +75,7 @@ def __init__(
         finished: bool,
         metrics: Optional[RequestMetrics] = None,
         lora_request: Optional[LoRARequest] = None,
+        embed = None
```
IMO it's a better idea to create a separate class for output of embedding models. Many attributes of `RequestOutput` are tied to language models but not relevant to embedding models. We can discuss this in the later stages of this PR

robertgshaw2-neuralmagic replied:
Agree @ywang96 

I think we will want to refactor `RequestOutput` and `SamplingParams` to be abstract with separate versions for `Completion` vs `Embedding`

It will be tricky to do this without breaking changes to the UX

robertgshaw2-neuralmagic commented:
@jc9123 any update?

jc9123 commented:
> @jc9123 any update?

Sorry for the delay but I will need a bit more time to finish milestone 1. I should be able to finish it by early next week.

brandonsorensen commented:
I don't want to step on @jc9123's toes, but I'd be happy to help out on this issue. For example, I could already start work on the front end bits.

jc9123 committed:
finished checkpoint 1.a

jc9123 committed:
cleaned up checkpoint 1.a

jc9123 commented:
> @jc9123 any update?

Hi Robert, sorry for the delay but I've finished checkpoint 1.a->1.d. Could you take a look; right now, I've mostly copied EmbeddingParams and EmbeddingOutput from their Sampling counterpart to ensure that everything still work but was wondering if there is something specific I should add/change.

jc9123 committed:
added embedded_model registry

jc9123 committed:
code cleanup & bugfix

jc9123 committed:
more cleanup

jc9123 committed:
more cleanup

CatherineSue commented:
@jc9123 @robertgshaw2-neuralmagic , it seems this PR has a few milestones that are overlapped with my PR https://github.com/vllm-project/vllm/pull/3734. I have also separated EmbeddingOutput from CompletionOutput. Should we collaborate so we don't do duplicate work?

jc9123 committed:
save

jc9123 committed:
fixed xlm_roberta

jc9123 commented:
> @jc9123 @robertgshaw2-neuralmagic , it seems this PR has a few milestones that are overlapped with my PR #3734. I have also separated EmbeddingOutput from CompletionOutput. Should we collaborate so we don't do duplicate work?

Hi Catherine, we should definitely collaborate to avoid duplicate work. Between our PR, I think we should have the same implementation for the vLLM front end UX and also the EmbeddingOutput and EmbeddingParams class.

jc9123 committed:
code cleanup

KylinMountain commented:
Any update? Can we use other embedding model now?

whybeyoung commented:
lgtm, but any update ?

simon-mo commented:
Bart has been added, cross attention as well, i believe there should be some small incrementally work in hooking this models with the embedding API 

If @jc9123 is not actively working on this anymore, please feel free to open a new PR base off this.

rpvelloso commented:
any updates? I'd like to use this model under openai API.

DarkLight1337 commented:
Closing as superseded by #3734 and #9387

Jay-ju commented:
During the inference process of bge-m3, testing seems to show no performance improvement.  Is this in line with expectations?

DarkLight1337 commented:
Most of vLLM's optimizations (e.g. chunked prefill, automatic prefix caching) focus on the generation process and therefore don't apply to embedding models. So, this is expected to some extent.
